src/tensors/cpu/prod_blas.h

```diff
@@ -39,7 +35,8 @@ inline void sgemm(bool transA,
                     beta,
                     c,
                     ldc);
-    #endif
+#elif USE_ONNX_SGEMM
+        gemm_f32_imp(transA, transB, rows_a, rows_b, width, alpha, a, b, beta, c);
```

Comment:
We shouldn't be using ONNX gemm. Instead, use the RUY gemm, and a postprocessing in the cases where the scaling factor is non-trivial. Wrap this around in a `RUY_GEMM_EXPANDED` function, or something similar

Reply:
Indeed I'm blocking the whole PR until RUY fp32 gemm is used.

Reply:
> We shouldn't be using ONNX gemm. Instead, use the RUY gemm, and a postprocessing in the cases where the scaling factor is non-trivial. Wrap this around in a RUY_GEMM_EXPANDED function, or something similar

@XapaJIaMnu can you take a glance at https://github.com/jerinphilip/sgemm/blob/1f7700a789691ec1202a599632b7c4b207b7328d/src/gemm-impl.cpp#L159-L244? I am not currently choosing to bring it here because testing along with marian's stuff is more a pain. At some point we are looking at integrating the implementation after benchmarking.

For the record, I disagree with you both regarding the strong push for ruy sgemm in this PR. There is a perfectly sane development trajectory where I don't add a new feature (sgemm using ruy). This first PR can just set the integration wiring in place using onnx sgemm (Eigen) - minimal diff and a followup PR can improve adding ruy sgemm. I am wary of checking in code without integration tests and the demand for an additional feature is also a demand for the tests. I would also like to note that the current changes to bring in ruy are at 500+ lines of change, so you're also looking at reviewing the same - @XapaJIaMnu has in the past complained about big PRs. 

Of course, I'm still trying to make progress in ruy implementation as instructed. The `CMakeLists.txt` and regression-tests infrastructure I have inherited from my predecessors and collaborators in this repository is not perhaps most ideal, which I hope everyone keeps in mind while fixating on speed.

Reply:
The gemm implementation that you proposed looks sufficient, even if it's a bit more complicated than what I would have gone for.

I'm happy to have it added to the pr. The batched implementation is optional as far as I recall, since it's only present in mkl and not in other Blas providers.

There are several merits to using ruy based gemm here.
1) You skip an eigen dependency which means it's much easier to eventually port to upstream (where eigen won't be accepted)
2) eigen is not particularly optimised for the type of work we do here. It's slow.
3) it's not that much extra work to get the gemm right.
4) you already have end to end testing
You will know when something stops work. Furthermore,  comparing eigen gemm vs your hand written ruy gemm would be trivial to do in a separate source file.

Reply:
I've changed to the title of the PR to reflect the original requirement which has been repeated a few times above: ruy for fp32 and int8.